.github/workflows/ci.yml

```diff
@@ -179,31 +186,159 @@ jobs:
+      - uses: 'actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c' # ratchet:actions/checkout@v3
+      - id: 'auth'
+        uses: 'google-github-actions/auth@ef5d53e30bbcd8d0836f4288f5e50ff3e086997d' # ratchet:google-github-actions/auth@v1
+        with:
+          workload_identity_provider: '${{ env.WIF_PROVIDER }}'
+          service_account: '${{ env.WIF_SERVICE_ACCOUNT }}'
+          token_format: 'access_token'
+      # Write the prefix into GITHUB_ENV which will be used in integration test as a prefix of the
+      # uploaded file's path.
+      - name: 'Write object prefix'
+        run: |
+          echo "INTEG_TEST_OBJECT_PREFIX=${{ needs.set_up_integration_infra.outputs.INTEG_TEST_OBJECT_PREFIX_OUTPUT }}" >> $GITHUB_ENV
+      # Write workflow id to GITHUB_ENV. This variable works as the unique identifier
+      # of the files uploaded by reusable workflow.
+      - name: 'write workflow id to env'
+        id: 'write_workflow_id'
+        run: |
+          echo "WORKFLOW_RUN_ID=${{ github.run_id }}" >> $GITHUB_ENV
       - name: 'Run tests'
         env:
           TEST_INTEGRATION: 'true'
         run: |-
           go test ./test/integration \
             -timeout=15m
+
+  clean_up_integration_infra:
+    runs-on: 'ubuntu-latest'
+    if: '${{ always() }}'
+    needs: ['set_up_integration_infra', 'run_integ_test']
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+    env:
+      TF_MODULE_DIR: './terraform/modules/pmap-service'
+      MAPPING_TF_MODULE_DIR: '/tmp/mapping_tf_module'
+      POLICY_TF_MODULE_DIR: '/tmp/policy_tf_module'
+      MAPPING_TF_VAR_FILE: '/tmp/mapping_tf_module/mapping_ci.tfvars'
+      POLICY_TF_VAR_FILE: '/tmp/policy_tf_module/policy_ci.tfvars'
+    steps:
+      - uses: 'actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c' # ratchet:actions/checkout@v3
+      - id: 'auth'
+        uses: 'google-github-actions/auth@ef5d53e30bbcd8d0836f4288f5e50ff3e086997d' # ratchet:google-github-actions/auth@v1
+        with:
+          workload_identity_provider: '${{ env.WIF_PROVIDER }}'
+          service_account: '${{ env.WIF_SERVICE_ACCOUNT }}'
+          token_format: 'access_token'
+      # TODO: Workload Identity doesn't work with GCS Provider.
+      # https://github.com/google-github-actions/auth/issues/89
+      - name: 'Workaround Terraform issue'
+        shell: 'bash'
+        run: |-
+          echo "GOOGLE_OAUTH_ACCESS_TOKEN=${{ steps.auth.outputs.access_token }}" >> $GITHUB_ENV
+      - name: 'Init the mapping terraform infrastructure for clean up'
+        if: '${{ always() }}'
+        run: |-
+          cp -r ${{ env.TF_MODULE_DIR }} ${{ env.MAPPING_TF_MODULE_DIR }}
+          terraform -chdir=${{ env.MAPPING_TF_MODULE_DIR }} init
+      - name: 'download artifact for mapping'
+        if: '${{ always() }}'
+        uses: 'actions/download-artifact@v3'
+        with:
+          name: 'terraform-mapping-files'
+          path: '${{ env.MAPPING_TF_MODULE_DIR }}'
```

Comment:
How do you ensure it is downloading the right files not the files uploaded by a different workflow run?

Reply:
I will add a random postfix to ensure that.